File: library/src/main/java/cn/wandersnail/widget/textview/RotatableTextView.java
package cn.wandersnail.widget.textview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.view.Gravity;

import androidx.appcompat.widget.AppCompatTextView;

import cn.wandersnail.widget.R;

/**
 * 可旋转的TextView
 * <p>
 * date: 2019/8/22 23:04
 * author: zengfansheng
 */
public class RotatableTextView extends AppCompatTextView {
    private float degree;

    public RotatableTextView(Context context) {
        super(context);
        init(null);
    }

    public RotatableTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context.obtainStyledAttributes(attrs, R.styleable.RotatableTextView));
    }

    public RotatableTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context.obtainStyledAttributes(attrs, R.styleable.RotatableTextView, defStyleAttr, 0));
    }

    private void init(TypedArray a) {
        setGravity(Gravity.CENTER);
        if (a != null) {
            degree = a.getFloat(R.styleable.RotatableTextView_wswDegree, degree);
            a.recycle();
        }
    }

    public float getDegree() {
        return degree;
    }

    public void setDegree(float degree) {
        this.degree = degree;
        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(getMeasuredWidth(), getMeasuredWidth());
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.save();
        canvas.translate(getCompoundPaddingLeft(), getExtendedPaddingTop());
        canvas.rotate(degree, getWidth() / 2f, getHeight() / 2f);
        super.onDraw(canvas);
        canvas.restore();
    }
}

File: library/src/main/java/com/snail/widget/textview/RotatableTextView.kt
package com.snail.widget.textview

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.view.Gravity
import androidx.appcompat.widget.AppCompatTextView
import com.snail.widget.R

/**
 * 可旋转的TextView
 *
 * date: 2018/12/29 14:21
 * author: zengfansheng
 */
open class RotatableTextView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = android.R.attr.textViewStyle) :
        AppCompatTextView(context, attrs, defStyleAttr) {
    var degree: Float = 0f
        set(value) {
            field = value
            invalidate()
        }
    
    init {
        gravity = Gravity.CENTER
        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.RotatableTextView)
            degree = a.getFloat(R.styleable.RotatableTextView_wswDegree, degree)
            a.recycle()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(measuredWidth, measuredWidth)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.save()
        canvas.translate(compoundPaddingLeft.toFloat(), extendedPaddingTop.toFloat())
        canvas.rotate(degree, this.width / 2f, this.height / 2f)
        super.onDraw(canvas)
        canvas.restore()
    }
}